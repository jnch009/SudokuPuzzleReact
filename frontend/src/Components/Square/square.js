import React from 'react';
import { FormInput } from 'shards-react';
import fn from '../../helperFn/boardFunctions';
import './square.scss';
import boxLookupAssignment from './squareUtility';
import tl from './squareBorders/topLeft';
import tr from './squareBorders/topRight';
import t from './squareBorders/Top';
import bl from './squareBorders/bottomLeft';
import br from './squareBorders/bottomRight';
import b from './squareBorders/Bottom';
import l from './squareBorders/Left';
import r from './squareBorders/Right';
import none from './squareBorders/None';
import squareBtnLight from './squareState/Active';
import squareBtnDanger from './squareState/Danger';
import disabledBtn from './squareState/Disabled';

const initialSquare = {
  edit: false,
  valid: true,
};

const rowTopLeft = [0, 3, 6];
const topRightRange = [2, 5, 8];
const topBorder = [1, 4, 7];

const boxConditionLookup = disabled => {
  const boxLookup = {};

  boxLookupAssignment(
    boxLookup,
    rowTopLeft,
    rowTopLeft,
    disabled,
    tl.SquareBorderTopLeft,
    tl.DisabledTopLeft,
  );

  boxLookupAssignment(
    boxLookup,
    rowTopLeft,
    topRightRange,
    disabled,
    tr.SquareBorderTopRight,
    tr.DisabledTopRight,
  );

  boxLookupAssignment(
    boxLookup,
    topRightRange,
    rowTopLeft,
    disabled,
    bl.SquareBorderBottomLeft,
    bl.DisabledBottomLeft,
  );

  boxLookupAssignment(
    boxLookup,
    topRightRange,
    topRightRange,
    disabled,
    br.SquareBorderBottomRight,
    br.DisabledBottomRight,
  );

  boxLookupAssignment(
    boxLookup,
    rowTopLeft,
    topBorder,
    disabled,
    t.SquareBorderTop,
    t.DisabledTop,
  );

  boxLookupAssignment(
    boxLookup,
    topRightRange,
    topBorder,
    disabled,
    b.SquareBorderBottom,
    b.DisabledBottom,
  );

  boxLookupAssignment(
    boxLookup,
    topBorder,
    rowTopLeft,
    disabled,
    l.SquareBorderLeft,
    l.DisabledLeft,
  );

  boxLookupAssignment(
    boxLookup,
    topBorder,
    topRightRange,
    disabled,
    r.SquareBorderRight,
    r.DisabledRight,
  );

  return boxLookup;
};

const borderColorLookup = (row, col) => {
  return {
    [`0 0`]: 'blue',
    [`3 3`]: 'blue',
    [`6 6`]: 'blue',
    [`0 3`]: 'yellow',
    [`3 6`]: 'yellow',
    [`6 0`]: 'yellow',
  }[`${row} ${col}`];
};

const setBorderColor = (row, col) => {
  //box 1,5 or 9
  return borderColorLookup(row, col) || '#c700ff';
};

let btn;

class Square extends React.Component {
  constructor(props) {
    super(props);
    this.state = initialSquare;
  }

  handleKeyPress = async e => {
    const { row, col, grid } = this.props;
    const number = e.key;

    //debugger;
    if (Number(number) !== Number(grid[row][col])) {
      if (!fn.isValid(grid, row, col, number)) {
        this.setState({ valid: false });
      } else {
        this.setState({ valid: true });
      }
    }

    if (e.key === 'Backspace' || e.key === 'Delete') {
      this.props.pressKey(null, row, col);
    } else {
      this.props.pressKey(e.key, row, col);
      this.handleClick();
    }
  };

  handleClick = () => {
    this.setState(() => ({ edit: !this.state.edit }));
  };

  render() {
    const { row, col, number } = this.props;
    const startRow = parseInt(row / 3) * 3;
    const startCol = parseInt(col / 3) * 3;
    const boxColor = setBorderColor(startRow, startCol);
    const boxLookup = boxConditionLookup()[`${row} ${col}`] || none.noBorder;
    const boxLookupDisabled =
      boxConditionLookup(true)[`${row} ${col}`] || none.DisabledNoBorder;

    if (this.props.modify === false) {
      btn = disabledBtn(boxLookupDisabled, number, boxColor);
    } else {
      if (!this.state.edit) {
        if (!this.state.valid && this.props.number !== null) {
          btn = squareBtnDanger(boxLookup, number, this.handleClick, boxColor);
        } else {
          btn = squareBtnLight(boxLookup, number, this.handleClick, boxColor);
        }
      } else {
        btn = (
          <FormInput
            autoFocus={true}
            onBlur={this.handleClick}
            type='text'
            pattern='[0-9]*'
            inputMode='numeric'
            onKeyDown={this.handleKeyPress}
            className='square'
            value={this.props.number !== null ? this.props.number : ''}
          ></FormInput>
        );
      }
    }

    return btn;
  }
}

export default Square;
