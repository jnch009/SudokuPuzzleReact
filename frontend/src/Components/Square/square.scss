$primary: #212529;
$danger: red;
$light: #fff;

.primary{
    background-color: $primary;
}

.danger{
    background-color: $danger;
}

.light{
    background-color: $light;
    transition: all 250ms cubic-bezier(.27,.01,.38,1.06);
}

.light:hover{
    color: #212529;
    background-color: #dadfe4;
    border-color: #dadfe4;
    box-shadow: 0 5px 15px rgba(0,0,0,.05), 0 4px 10px rgba(233,236,239,.25);
}