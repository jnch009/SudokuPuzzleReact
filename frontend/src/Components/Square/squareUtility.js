const boxLookupAssignment = (boxLookup, rowArr, colArr, disabled, buttonNonDisabled, buttonDisabled) => {
  for (let rowNum = 0; rowNum < rowArr.length; rowNum++) {
    for (let colNum = 0; colNum < colArr.length; colNum++) {
      boxLookup[`${rowArr[rowNum]} ${colArr[colNum]}`] = !disabled
        ? buttonNonDisabled
        : buttonDisabled;
    }
  }
};

export default boxLookupAssignment;
