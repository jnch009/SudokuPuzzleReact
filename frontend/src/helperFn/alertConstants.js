export const alertTypes = {
  SUCCESS: 'success',
  ERROR: 'danger',
}