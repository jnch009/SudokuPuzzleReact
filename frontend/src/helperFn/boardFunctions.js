function randomlyGeneratedValue(min, max) {
  return Math.floor(Math.random() * (max - min)) + min;
}

const removingEntries = (newGrid, difficulty) => {
  let minToRemove;
  let maxToRemove;
  if (difficulty === 'Beginner') {
    minToRemove = 1;
    maxToRemove = 1;
  } else if (difficulty === 'Easy') {
    minToRemove = 3;
    maxToRemove = 3;
  } else if (difficulty === 'Normal') {
    minToRemove = 4;
    maxToRemove = 6;
  } else if (difficulty === 'Hard') {
    minToRemove = 7;
    maxToRemove = 10;
  }

  for (let row = 0; row < newGrid.length; row += 1) {
    let entriesToRemove = randomlyGeneratedValue(minToRemove, maxToRemove);
    let entriesRemoved = 0;
    let indexEntries = [0, 1, 2, 3, 4, 5, 6, 7, 8];

    while (entriesRemoved < entriesToRemove) {
      let entryRemoved = randomlyGeneratedValue(0, indexEntries.length);
      newGrid[row].splice(indexEntries[entryRemoved], 1, null);
      indexEntries.splice(entryRemoved, 1);
      entriesRemoved += 1;
    }
  }
};

const createGrid = () => {
  let arr = [];

  for (let i = 0; i < 9; i++) {
    let innerArr = [];
    for (let j = 0; j < 9; j++) {
      innerArr.push(null);
    }
    arr.push(innerArr);
  }

  return arr;
};

// I could be wrong here, but we only need to check one condition
// If this condition fails, then they all fail
const verifyRow = (grid, row, val = null) => {
  const rowOfNums = Array.from(grid[row], (x) => Number(x));
  if (!val) {
    let verifyArr = [];
    rowOfNums.forEach((row) => {
      if (!verifyArr.includes(row)) {
        verifyArr.push(row);
      }
    });
    return verifyArr.length === grid[row].length;
  } else {
    if (rowOfNums.includes(Number(val))) {
      return false;
    }
    return true;
  }
};

const verifyCol = (grid, col, val = null) => {
  for (let row = 0; row < 9; row++) {
    if (Number(grid[row][col]) === Number(val)) {
      return false;
    }
  }
  return true;
};

const verifyBox = (grid, row, col, val = null) => {
  let startRow = parseInt(row / 3) * 3;
  let startCol = parseInt(col / 3) * 3;

  for (let r = startRow; r < startRow + 3; r++) {
    for (let c = startCol; c < startCol + 3; c++) {
      if (Number(grid[r][c]) === Number(val)) {
        return false;
      }
    }
  }
  return true;
};

const isValid = (grid, row, col, num = null) => {
  if (
    verifyBox(grid, row, col, num) && verifyCol(grid, col, num) && verifyRow(grid, row, num)
  ) {
    return true;
  }
  return false;
};

const verifySudoku = (grid) => {
  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      if (!isValid(grid, row, col)) {
        return false;
      }
    }
  }
  return true;
};

// https://stackoverflow.com/questions/6274339/how-can-i-shuffle-an-array fisher yates
const shuffle = (a) => {
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
};

const solve = (grid, shuffled) => {
  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      if (grid[row][col] === null) {
        shuffled.forEach((choice) => {
          if (isValid(grid, row, col, choice)) {
            grid[row][col] = choice;
            solve(grid, shuffled);
            // this is something that I added to stop the recursion when a solution is found otherwise it finds every solution!
            if (!verifySudoku(grid)) {
              grid[row][col] = null;
            }
          }
        });
        return;
      }
    }
  }
  return grid;
};

export default {
  randomlyGeneratedValue,
  verifyRow,
  verifyCol,
  verifyBox,
  removingEntries,
  createGrid,
  isValid,
  verifySudoku,
  shuffle,
  solve,
};
