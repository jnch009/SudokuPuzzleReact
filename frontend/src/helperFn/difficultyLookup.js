const difficulties = ['Beginner','Easy','Normal','Hard'];
export default difficulties;
