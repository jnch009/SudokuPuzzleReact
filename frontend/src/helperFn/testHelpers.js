const squareStateConstants = {
  num: 1,
  handleClick: () => {},
  borderColor: 'blue'
};

export {
  squareStateConstants,
};
