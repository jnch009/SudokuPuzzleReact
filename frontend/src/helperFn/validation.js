//validating the save game name on the frontend
const regex = new RegExp(process.env.REACT_APP_REGEX_CURSE_WORDS);
const validateSaveName = (name) => {
  if (name.length === 0 || name.length > 100 || regex.test(name.toLowerCase())) {
    return false;
  }

  return true;
};

export default validateSaveName;
