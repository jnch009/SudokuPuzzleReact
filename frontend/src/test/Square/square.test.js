import React from 'react';
import 'jest-styled-components';

import { render, screen } from '@testing-library/react';
import { squarePropsConstants } from '../../helperFn/testHelpers';
import Square from '../../Components/Square/square';

//integration test
const borderColor = 'blue';

it('Valid Square', () => {
  const { container } = render(<Square squareProps={squarePropsConstants} modify={true} valid={true} />);
  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild).toHaveStyleRule('border-top', `0.2em ${borderColor} solid`);
  expect(container.firstChild).toHaveStyleRule('border-left', `0.2em ${borderColor} solid`);
  expect(container.firstChild.classList.contains('square')).toBeTruthy();
  expect(container.firstChild.classList.contains('light')).toBeTruthy();
  expect(screen.getByText('1')).toBeTruthy();
});

