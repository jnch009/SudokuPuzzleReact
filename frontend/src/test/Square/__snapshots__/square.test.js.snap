// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Valid Square 1`] = `
.c0 {
  border-top: 0.2em blue solid;
  border-left: 0.2em blue solid;
}

<button
  class="c0 square light"
>
  1
</button>
`;
